### PR TITLE
Replace schema SQL collectors with SQL builders

### PR DESCRIPTION
**Pull request by morozov**
The approach of using the Visitor pattern for building SQL is wrong. This pattern is useful when the visitor knows nothing about the internal structure of the object it visits, and the object being visited is responsible for walking the visitor through its internals.

It doesn't apply to building SQL for schema objects because:

1. The internal structure of the schema objects is known and is always the same.
2. The actual logic of building SQL depends on the specific target platform, not the abstract schema.

More specifically, in order to enable the support for foreign keys on SQLite, we need the logic of building the `CREATE TABLE` and `DROP TABLE` statements to be platform-specific when it comes to the tables that reference each other via foreign keys:

1. All platforms except for SQLite require the referenced table to exist at the time of the creation of the foreign key. So in the case of circular references, tables need to be created without some foreign keys and then foreign keys need to be created separately. This is how it currently works.
2. SQLite doesn't support adding or dropping foreign keys to an existing table but it doesn't require the referenced table to exist at the time of the creation of the foreign key (the constraint is enforced during row modification). Therefore, we need to create each table with its foreign keys in a single statement.

The above is impossible to implement via the Visitor API due to the reasons described in https://github.com/doctrine/dbal/pull/5409#issuecomment-1135237936.

### CODE REVIEW

File: src/Platforms/AbstractMySQLPlatform.php [hunk 1]
```diff
@@ -407,6 +408,38 @@ public function getListTableMetadataSQL(string $table, ?string $database = null)
...
+                foreach ($table->getForeignKeys() as $foreignKey) {
+                    $sql[] = $this->getCreateForeignKeySQL(
+                        $foreignKey,
+                        $table->getQuotedName($this)
+                    );
+                }
+            } elseif (count($table->getForeignKeys()) > 0) {
+                Deprecation::trigger(
+                    'doctrine/dbal',
+                    'https://github.com/doctrine/dbal/pulls/5414',
```
Comment (greg0ire) [inline src/Platforms/AbstractMySQLPlatform.php]:
```suggestion
                    'https://github.com/doctrine/dbal/pull/5414',
```

File: src/Schema/Visitor/CreateSchemaSqlCollector.php [hunk 2]
```diff
@@ -28,6 +32,12 @@ class CreateSchemaSqlCollector extends AbstractVisitor
 
     public function __construct(AbstractPlatform $platform)
     {
+        Deprecation::trigger(
+            'doctrine/dbal',
+            'https://github.com/doctrine/dbal/pulls/5416',
```
Comment (greg0ire) [inline src/Schema/Visitor/CreateSchemaSqlCollector.php]:
```suggestion
            'https://github.com/doctrine/dbal/pull/5416',
```

File: src/Schema/Visitor/DropSchemaSqlCollector.php [hunk 3]
```diff
@@ -31,6 +34,12 @@ class DropSchemaSqlCollector extends AbstractVisitor
 
     public function __construct(AbstractPlatform $platform)
     {
+        Deprecation::trigger(
+            'doctrine/dbal',
+            'https://github.com/doctrine/dbal/pulls/5416',
```
Comment (greg0ire) [inline src/Schema/Visitor/DropSchemaSqlCollector.php]:
```suggestion
            'https://github.com/doctrine/dbal/pull/5416',
```

File: tests/Schema/SchemaDiffTest.php [hunk 4]
```diff
@@ -1,134 +0,0 @@
...
-use Doctrine\DBAL\Platforms\AbstractPlatform;
-use Doctrine\DBAL\Schema\ForeignKeyConstraint;
-use Doctrine\DBAL\Schema\SchemaDiff;
-use Doctrine\DBAL\Schema\Sequence;
-use Doctrine\DBAL\Schema\Table;
-use Doctrine\DBAL\Schema\TableDiff;
-use PHPUnit\Framework\MockObject\MockObject;
-use PHPUnit\Framework\TestCase;
-
-class SchemaDiffTest extends TestCase
```
Comment (greg0ire) [inline tests/Schema/SchemaDiffTest.php:14]: Why are these tests removed?

Comment (morozov) [reply tests/Schema/SchemaDiffTest.php:14]: These tests mock the methods that are no longer used for schema generation (e.g. `getCreateForeignKeySQL ()`). The tests that make assertions against SQL provide more harm than good, so it is not worth updating them according to the code changes.

There's a new integration test (`CreateAndDropSchemaObjectsSQLBuilderTest`) which covers basic scenarios.

**Review by greg0ire — APPROVED** [review]

**Pull request merged**